runtime/buffer.h

```diff
@@ -242,10 +242,12 @@ class raw_buffer {
     for (std::size_t i = slice_leading; i < ds.size(); ++i) {
       std::size_t d = ds[i];
       std::size_t next_d = i + 1 < ds.size() ? ds[i + 1] : rank;
+      assert(d < rank);
+      assert(next_d <= rank);
 
       // Move the dimensions between this slice and the next slice down by the number of slices we've done so far.
       for (std::size_t j = d; j + 1 < next_d; ++j) {
-        dims[j] = dims[j + i + 1];
+        dims[j] = dims[j + 1];
```

Comment:
This may have been buggy but I still think something like the old code was necessary. I think you should add a test that slices dimensions 0, 2, and 4 from a 6 dimensional buffer. Wouldn't that test fail without this?